### ISSUE TITLE
[CURATOR-7] Session ids not preserved if EnsembleProvider has changed the ensemble

### ISSUE THREAD

**Issue by jira-importer** [apache/curator#528]
<p>See <a href="https://github.com/Netflix/curator/issues/266" class="external-link" target="_blank" rel="nofollow noopener">https://github.com/Netflix/curator/issues/266</a></p>

<p>InterProcessMutex, LeaderLatch, etc use an ephemeral node. If ZooKeeper gives a Disconnected event, the native client reconnects with the same session id, and the ephemeral node is preserved. If the ensemble changes at any point before a Disconnect, Curator's ConnectionState#checkState() calls handleNewConnectionString(), which constructs a new native ZK client, discarding the previous session id, and losing all locks.</p>

<p>This can be expensive.</p>

<p>Can ConnectionState be made to preserve the session id, or be more conservative about discarding the entire native client on a Disconnected event?</p>

<p>Clarifications:</p>

<p>An ensemble change (e.g. adding a new node to a cluster) does not mean that all session ids are now invalid. The next network glitch should not break locks.</p>

<p>There is no assumption in this description that the reconfiguration and the glitch/disconnection were related - they need not be.</p>

---
<details><summary><i>Originally reported by <a title="Shevek" href="https://issues.apache.org/jira/secure/ViewProfile.jspa?name=arren">arren</a>, imported from: <a href="https://issues.apache.org/jira/browse/CURATOR-7" target="_blank">Session ids not preserved if EnsembleProvider has changed the ensemble</a></i></summary>
<i><ul>
<li><b>assignee</b>: <a title="Jordan Zimmerman" href="https://issues.apache.org/jira/secure/ViewProfile.jspa?name=randgalt">randgalt</a>
<li><b>status</b>: Closed
<li><b>priority</b>: Major
<li><b>resolution</b>: Implemented
<li><b>resolved</b>: 2023-05-10T13:46:41+01:00
<li><b>imported</b>: 2025-01-21
</ul></i>
</details>

**Comment by jira-importer** [comment]
<i><a href="https://issues.apache.org/jira/secure/ViewProfile.jspa?name=randgalt">randgalt</a>:</i>
<p>Possible implementations (off the top of my head): HandleHolder could be changed to hold the session info. It would get assigned by ConnectionState when the session is established. ZookeeperFactory needs additional arguments for optional session info. If the HandleHolder has session info, it passes it to the ZookeeperFactory. ConnectionState would need to handle KeeperState.Expired in this case (though it may just handle it normally - you'd need to test for that).</p>

**Comment by jira-importer** [comment]
<i><a href="https://issues.apache.org/jira/secure/ViewProfile.jspa?name=arren">arren</a>:</i>
<p>The ZooKeeper documentation suggests that sessionId may change after a reconnect (within the native ZK handle) so I suspect we can safely pass a duff sessionId through ZookeeperFactory.</p>

**Comment by jira-importer** [comment]
<i><a href="https://issues.apache.org/jira/secure/ViewProfile.jspa?name=randgalt">randgalt</a>:</i>
<p>I don't think so. If I remember correctly that will generate an error. If there is no session info (you need both ID and password, I believe) then use the ZK constructor that doesn't take those. </p>

**Comment by jira-importer** [comment]
<i><a href="https://issues.apache.org/jira/secure/ViewProfile.jspa?name=arren">arren</a>:</i>
<p>Did ZooKeeper have the HostProvider interface at the point Curator Ensemble was written? Is there a reason why Ensemble isn't implemented in terms of a HostProvider?</p>

**Comment by jira-importer** [comment]
<i><a href="https://issues.apache.org/jira/secure/ViewProfile.jspa?name=arren">arren</a>:</i>
<p>It looks as if one HAS to do this in ConnectionState/HandleHolder, because ConnectionStateManager delivers ConnectionState (the OTHER ConnectionState) events on a separate ExecutorService thread, so anything trying to compute validity of the ZooKeeper handle externally is subject to a race condition.</p>

**Comment by jira-importer** [comment]
<i><a href="https://issues.apache.org/jira/secure/ViewProfile.jspa?name=arren">arren</a>:</i>
<p>ZooKeeper ClientCnxn line 850:     long sessId = (seenRwServerBefore) ? sessionId : 0;</p>

<p>The only person who sets seenRwServerBefore is onConnected().</p>

<p>Therefore, it appears that passing a sessionId into a ZooKeeper constructor has no effect, as the ClientCnxn has never seen an RW server before, so it discards it anyway. So the ZookeeperFactory approach won't work.</p>

**Comment by jira-importer** [comment]
<i><a href="https://issues.apache.org/jira/secure/ViewProfile.jspa?name=randgalt">randgalt</a>:</i>
<p>That does seem strange. Maybe bring that up on the ZK dev list? As coded, it seems like the passed in session ID will always be ignored. I need to study the code more though.</p>

**Comment by jira-importer** [comment]
<i><a href="https://issues.apache.org/jira/secure/ViewProfile.jspa?name=arren">arren</a>:</i>
<p>I am currently trying to code a patch against the ZK 3.5.0 API, which obviously can't be committed to curator yet, but I think we can (carefully) run the 3.5.x client against a 3.4 server for our purposes.</p>

<p>The relevant issue in ZooKeeper which allows this is <a href="https://issues.apache.org/jira/browse/ZOOKEEPER-1355" title="Add zk.updateServerList(newServerList) " class="issue-link" data-issue-key="ZOOKEEPER-1355"><del>ZOOKEEPER-1355</del></a>.</p>

**Comment by jira-importer** [comment]
<i><a href="https://issues.apache.org/jira/secure/ViewProfile.jspa?name=randgalt">randgalt</a>:</i>
<p>3.5 will be much simpler as it supports reconfiguring the client. You may not want to spend too much time 3.5 as I will need to rework a lot of the Exhibitor/Curator interaction in light of dynamic configuration.</p>

**Comment by jira-importer** [comment]
<i><a href="https://issues.apache.org/jira/secure/ViewProfile.jspa?name=arren">arren</a>:</i>
<p>Unfortunately I have a deadline to fix this, so unless there are any other suggestions, I need to do this today. I would very much appreciate the value of your review, but will not be at all offended if you don't want my code in the long run.</p>

**Comment by jira-importer** [comment]
<i><a href="https://issues.apache.org/jira/secure/ViewProfile.jspa?name=arren">arren</a>:</i>
<p>ConnectionState.checkTimeouts resets the connection if the session timeout has expired, whether or not we ever had a session. If one has a large, shooty cluster with a short session timeout and a long connection timeout, this might reset the host list pointer to zero every sessionTimeoutMs, never letting it connect to a valid, available host towards the end of the list. On the other hand, if it is an assertion that sessionTimeout > connectionTimeout always, then the min() call is not required and the ifelse should read "else if (elapsed > connectionTimeoutMs)" instead of just "else".</p>

**Comment by jira-importer** [comment]
<i><a href="https://issues.apache.org/jira/secure/ViewProfile.jspa?name=randgalt">randgalt</a>:</i>
<p>Actually, I thought of that same thing. That should be added. Please open a separate Issue for that.</p>

**Comment by jira-importer** [comment]
<i><a href="https://issues.apache.org/jira/secure/ViewProfile.jspa?name=arren">arren</a>:</i>
<p>It's not a beauty contest, but here's the code. I'd love feedback, and you have my blessing to do as you please with it.</p>

**Comment by jira-importer** [comment]
<i><a href="https://issues.apache.org/jira/secure/ViewProfile.jspa?name=randgalt">randgalt</a>:</i>
<p>I'm out of town until April 21st - I'll check when I get back.</p>

**Comment by jira-importer** [comment]
<i><a href="https://issues.apache.org/jira/secure/ViewProfile.jspa?name=arren">arren</a>:</i>
<p>See ZooKeeper-1683 for NPE in ZK itself.</p>

**Comment by jira-importer** [comment]
<i><a href="https://issues.apache.org/jira/secure/ViewProfile.jspa?name=mdrob">mdrob</a>:</i>
<p>This patch has been sitting in limbo for over a year now. Is it still an issue? If so, I can do the housekeeping necessary on the patch.</p>

**Comment by jira-importer** [comment]
<i><a href="https://issues.apache.org/jira/secure/ViewProfile.jspa?name=randgalt">randgalt</a>:</i>
<p>We're trying to keep up. For most of Curator's life I was the only active committer. We're get more committers, but any help you can provide would be appreciated.</p>

**Comment by jira-importer** [comment]
<i><a href="https://issues.apache.org/jira/secure/ViewProfile.jspa?name=mdrob">mdrob</a>:</i>
<p>Attaching a version of Shevek's patch against the apache code base. It looks like it depends on ZK-3.5, based on his original naming, so it does not compile as applied to master. I have not looked at it in enough detail to figure out if that is something that can be aligned with ZK-3.4 or not.</p>
